src/api/BundleDeploy/BundleDeployer.ts

```diff
@@ -0,0 +1,97 @@
+/*
+* This program and the accompanying materials are made available under the terms of the
+* Eclipse Public License v2.0 which accompanies this distribution, and is available at
+* https://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Copyright IBM Corp, 2019
+*
+*/
+
+"use strict";
+
+import { IHandlerParameters, Logger, ImperativeError } from "@brightside/imperative";
+import { ZosmfSession, SubmitJobs, List } from "@brightside/core";
+import { ParmValidator } from "./ParmValidator";
+
+/**
+ * Class to represent a CICS Bundle Deployer.
+ *
+ * @export
+ * @class BundleDeployer
+ */
+export class BundleDeployer {
+
+  private params: IHandlerParameters;
+
+  /**
+   * Constructor for a BundleDeployer.
+   * @param {IHandlerParameters} params - The Imperative handler parameters
+   * @throws ImperativeError
+   * @memberof BundleDeployer
+   */
+  constructor(params: IHandlerParameters) {
+    this.params = params;
+  }
+
+  /**
+   * Deploy a CICS Bundle
+   * @returns {Promise<string>}
+   * @throws ImperativeError
+   * @memberof BundleDeployer
+   */
+  public async deployBundle(): Promise<string> {
+
+    // Validate that the parms are valid for Deploy
+    ParmValidator.validateDeploy(this.params);
+
+    // Create a zosMF session
+    let zosmfProfile;
+    try {
+      zosmfProfile = this.params.profiles.get("zosmf");
+    }
+    catch (error) {
+      // Unit tests tend to cause an exception here
+      throw new Error("No zosmf profile found");
```

Comment:
If this is just here for the unit tests then I think you could get around this by mocking `profiles.get` in the

Reply:
I wanted to use an object literal function definition... but it wouldn't work. It either throws an exception at runtime, or a typescript error at compile time. I thought it might be neat to embed a template in the data that could be returned by the get() function (the tests for the sample project imply that this should be possible), but after a whole load of unhelpful error messages I got fed up with trying.

Reply:
I've kinda adopted this, but not exactly. The problem is that removing the deep copy of DEFAULT_PARAMETERS means that every time we set a parameter, it stays set for the subsequent tests. I've restructured the tests so that we not only set the parms we want to have set, but unset those that shouldn't be set, that avoids the need for the deep copy and allows us to revert back to the approach recommended here.